spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/BinderFactoryWithHealthIndicator.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder;
+
+import java.util.Map;
+
+import org.springframework.boot.actuate.health.AbstractHealthIndicator;
+import org.springframework.boot.actuate.health.CompositeHealthIndicator;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.boot.actuate.health.OrderedHealthAggregator;
+import org.springframework.context.ConfigurableApplicationContext;
+
+/**
+ * {@link BinderFactory} that extends {@link DefaultBinderFactory} and provides {@link HealthIndicator} support.
+ *
+ * @author Ilayaperumal Gopinathan
+ */
+public class BinderFactoryWithHealthIndicator extends DefaultBinderFactory {
+
+	private volatile CompositeHealthIndicator bindersHealthIndicator;
+
+	public BinderFactoryWithHealthIndicator(Map<String, BinderConfiguration> binderConfigurations) {
+		super(binderConfigurations);
+	}
+
+	public void setBindersHealthIndicator(CompositeHealthIndicator bindersHealthIndicator) {
+		this.bindersHealthIndicator = bindersHealthIndicator;
+	}
+
+	@Override
+	protected <T> Binder<T, ?, ?> getBinderInstance(String binderConfigurationName) {
+		Binder<T, ?, ?> binder = super.getBinderInstance(binderConfigurationName);
+		if (this.bindersHealthIndicator != null) {
```

Comment:
Can we move this into a 'Listener' type of class registered with the factory - maybe add other types of customizable behaviour in the future. Subclassing feels very invasive.